### ISSUE TITLE
Missing Arrow Integration Types

### ISSUE THREAD

**Issue by pdet** [duckdb/duckdb#1742]
The idea is to use this task to keep track of missing Arrow types in our integration.
 **If you are an arrow-duckdb user**, please speak up here if you are missing any (or other) of the following types and I will take it into consideration when prioritizing them.

Missing Types: (Priority List)
- [x] 128 Bit decimal (PR: #1755)
- [x] List (Normal, Large, Fixed-Size  [123 items])
- [x] Struct
- [x] Map (#1466)
- [x] Dictionary (#1375)
- [x] Binary (Normal, Large)
- [x] Fixed-width binary [42 bytes]
- [x] Large UTF-8 String
- [x] Date 64 (Milisseconds)
- [x] Time32 (Milliseconds)
- [x] Time64 (Microseconds,Nanoseconds)
- [x] Duration (Milliseconds,Microseconds,Nanoseconds, Seconds)
- [x] Interval ([Months] [Day,Time]
- [ ] Half-Float
- [ ] Union (Sparse and Dense)
- [ ] 256 Bit Decimal

Missing Support:
- [x] Projection Pushdown
- [x] Avoid String Copy
- [x] FIlter Pushdown
- [x] Timestamp with timezones (e.g.,pa.timestamp(unit="ns", tz="UTC"))) (Depends on #561) 

Reference: https://arrow.apache.org/docs/format/CDataInterface.html?highlight=decimal

**Comment by akokkonidis** [comment]
perhaps [dictionary type](https://arrow.apache.org/docs/python/data.html#dictionary-arrays)?
As arrow uses dictionary types to improve performance & save memory on low-cardinality columns (similar to pandas.Categorical) it could be interesting for a lot of use cases
 
#https://github.com/duckdb/duckdb/issues/1375

**Comment by pdet** [comment]
Hi @akokkonidis, thanks for pointing out your interest, I've added it to (near the top) of the list, and will ping you whenever that's done :-)

**Comment by ldacey** [comment]
In addition to the dictionary type, I also ran into errors with timestamps with time zones. A lot of my data is saved with this type:

`pa.timestamp(unit="ns", tz="UTC"))`

**Comment by manifest** [comment]
It seems that `list_(uint8())` stored in a parquet file is being loaded as `BLOB` instead of `LIST<UTINYINT>`.

```python
import duckdb
import pyarrow as pa
from pyarrow import parquet as pq

def create_dataset(type):
  schema = pa.schema([("data", type)])
  inputs = [pa.array([b"foo", b"foobar", b"baz"], type=type)]
  table = pa.Table.from_arrays(inputs, schema=schema)
  pq.write_table(table, "/tmp/test.parquet")

con = duckdb.connect(database=':memory:', read_only=False)
```
```python
create_dataset(pa.list_(pa.uint8()))
con.execute("SELECT * FROM parquet_scan('/tmp/test.parquet')").df()
# RuntimeError: Unsupported type LIST<UTINYINT> for DuckDB -> NumPy conversion
con.execute("SELECT * FROM parquet_scan('/tmp/test.parquet')").arrow().to_pandas()
#                            data
# 0               [102, 111, 111]
# 1  [102, 111, 111, 98, 97, 114]
# 2                 [98, 97, 122]
con.execute("CREATE TABLE test AS SELECT * FROM parquet_scan('/tmp/test.parquet')")
# RuntimeError: TransactionContext Error: Failed to commit: Not implemented Error: Unimplemented type for uncompressed segment
con.execute("CREATE TABLE test (data BLOB)")
con.execute("INSERT INTO test SELECT CAST(data as BLOB) FROM parquet_scan('/tmp/test.parquet')")
# RuntimeError: Not implemented Error: Unimplemented type for cast (LIST<UTINYINT> -> BLOB)
con.execute("DROP TABLE test");
```

I'd also vote for support of casting between `LIST<UTINYINT>` and `BLOB`, and for support of Arrow's `binary()` data type to make the DuckDB's Arrow/Parquet API cosistent.

```python
create_dataset(pa.binary())
con.execute("SELECT * FROM parquet_scan('/tmp/test.parquet')").df()
#                            data
# 0               [102, 111, 111]
# 1  [102, 111, 111, 98, 97, 114]
# 2                 [98, 97, 122]
con.execute("SELECT * FROM parquet_scan('/tmp/test.parquet')").arrow()
# RuntimeError: Not implemented Error: Unsupported Arrow type BLOB
con.execute("CREATE TABLE test AS SELECT * FROM parquet_scan('/tmp/test.parquet')")
con.execute("SELECT * from test").df()
con.execute("SELECT * from test").array()
# TypeError: 'NoneType' object is not callable
```

Casting Arrow's `list_(uint8())` and `binary()` data types to `LIST<UTINYINT>` and `BLOB`, then to  NumPy's char arrays with `dtype=bytes` also seems resonable and may improve consistency of Arrow/NumPy API.

**Comment by pdet** [comment]
Hi @manifest, I think that is a good idea, I've brought the integration of binaries up on my priority list, thanks for bringing it up.

**Comment by pdet** [comment]
@ldacey We don't really have support internally to timezones yet, I'm wondering if your data is only UTC+/-x (e.g., UTC+1), I could support that by normalizing all columns to UTC internally. Would that be sufficient for you?
Note that this wouldn't round-trip though, as soon as you get your data from duckdb to arrow the timezone information would be lost.

**Comment by alanpaulkwan** [comment]
Will this support auto-casting the arrow byte binary to strings automatically? This is one pain right now.

**Comment by pdet** [comment]
Hi @alanpaulkwan could you send an example of your problem?
If I remember correctly the arrow binary types will be converted to DuckDB Blobs and casts from blob to varchar exist
e.g.,
```sql
SELECT b::VARCHAR FROM blobs 
```
I'm not so sure if there is implicit casting implemented, is that what you desire?

**Comment by alanpaulkwan** [comment]
@pdet the problem I'm facing is just when I output a file from Clickhouse into Parquet or Impala, sometimes what I know is a string is read in R as an arrow_binary. However, If I type CAST(variablename as VARCHAR) it reads correctly. The same happens with R's arrow package - is DuckDB using the same Parquet reader?

I thought about writing a workaround whereby I detect based on PARQUET_METADATA whether something is a binary or a string from the perspective of R and cast it when I read it in. But I don't know if DuckDB can tell the difference. When I tried this copy (select variable,cast(variable as varchar) variable_string ).... to a parquet file, and Parquet Metadata says they're both BYTE_BINARY. 

Ideally, there would be a flag I could turn on where we try to cast these types to strings without casting in case I do a lazy select *

**Comment by pdet** [comment]
Hi, @alanpaulkwan I'm a bit confused on how this relates to the arrow integration if it all seems to be a problem with the parquet reader?
Could you post the problematic code + parquet reader? (If it's sensitive data, feel free to email it to me holanda@cwi.nl) And I'm happy to take a look at it!

**Comment by alanpaulkwan** [comment]
@pdet thank you. I emailed it to you.

**Comment by Mytherin** [comment]
> @pdet the problem I'm facing is just when I output a file from Clickhouse into Parquet or Impala, sometimes what I know is a string is read in R as an arrow_binary. However, If I type CAST(variablename as VARCHAR) it reads correctly. The same happens with R's arrow package - is DuckDB using the same Parquet reader?
> 
> I thought about writing a workaround whereby I detect based on PARQUET_METADATA whether something is a binary or a string from the perspective of R and cast it when I read it in. But I don't know if DuckDB can tell the difference. When I tried this copy (select variable,cast(variable as varchar) variable_string ).... to a parquet file, and Parquet Metadata says they're both BYTE_BINARY.
> 
> Ideally, there would be a flag I could turn on where we try to cast these types to strings without casting in case I do a lazy select *

The problem here is likely that certain Parquet writers (especially older ones) do not correctly write the metadata of the file to specify that the columns are UTF8 strings, but instead write out the columns as blobs (and on import assume that all blobs are UTF8 strings). 

[Spark](https://spark.apache.org/docs/latest/sql-data-sources-parquet.html) has a flag to work around this problem in their parquet reader:

```
spark.sql.parquet.binaryAsString

Some other Parquet-producing systems, in particular Impala, Hive, and older versions of Spark SQL, do not differentiate between binary data and strings when writing out the Parquet schema. This flag tells Spark SQL to interpret binary data as a string to provide compatibility with these systems.	
```

We should add support for this flag as well.

**Comment by pdet** [comment]
Hi @alanpaulkwan, I've just checked your file, is exactly the situation mentioned by Mark. I'll add a flag to the parquet scan as mark suggested.

**Comment by alanpaulkwan** [comment]
Thank you. I have an inefficient workaround involving using DuckDB to
rewrite a copy of the file but this would I think be helpful to a number of
people. Thanks!


*Alan Kwan*

+852-9794-9483 (HK)

(626)-216-4899 (USA)

**Comment by holdenmatt** [comment]
I've been playing with DuckDB as a query engine on Arrow (pretty sweet!), but just hit this bug:
"Not implemented Error: Unsupported Internal Arrow Type tsn:UTC"

We are storing dates using this type: `pa.timestamp(unit="ns", tz="UTC"))`.
Works if I exclude the tz field, but it's a shame to only be able to use timezone naive types.

Looks like this is already tracked above, and depends on #561 (which has progressed recently).

Would be very helpful to support Arrow timezones here as well :)

**Comment by pdet** [comment]
@Mytherin, I guess we can probably close this?
I don't think we will be adding support for the other data types anytime soon.
